File: src/extras-kcoreaddons/libktexttohtml.cpp
#include <KTextToHTML>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <ktexttohtml.h>
#include "libktexttohtml.h"
#include "libktexttohtml.hxx"

libqt_string KTextToHTML_ConvertToHtml(const libqt_string param1, const int* param2, int param3, int param4) {
    QString param1_QString = QString::fromUtf8(param1.data, param1.len);
    QString _ret = KTextToHTML::convertToHtml(param1_QString, (const KTextToHTML::Options&)(*param2), static_cast<int>(param3), static_cast<int>(param4));
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

File: src_original/extras-kcoreaddons/libktexttohtml.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const ktexttohtml_enums = enums;
const std = @import("std");

/// https://api.kde.org/ktexttohtml.html
pub const ktexttohtml = struct {
    /// [Qt documentation](https://api.kde.org/ktexttohtml.html#convertToHtml)
    ///
    /// ``` param1: []const u8, param2: *const flag of ktexttohtml_enums.Option, param3: i32, param4: i32, allocator: std.mem.Allocator ```
    pub fn ConvertToHtml(param1: []const u8, param2: *const i32, param3: i32, param4: i32, allocator: std.mem.Allocator) []const u8 {
        const param1_str = qtc.libqt_string{
            .len = param1.len,
            .data = param1.ptr,
        };
        const _str = qtc.KTextToHTML_ConvertToHtml(param1_str, @ptrCast(param2), @intCast(param3), @intCast(param4));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("ktexttohtml.ConvertToHtml: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }
};

/// https://api.kde.org/ktexttohtml.html#types
pub const enums = struct {
    pub const Option = enum {
        pub const PreserveSpaces: i32 = 2;
        pub const ReplaceSmileys: i32 = 4;
        pub const IgnoreUrls: i32 = 8;
        pub const HighlightText: i32 = 16;
        pub const ConvertPhoneNumbers: i32 = 32;
    };
};
